util/run.bzl

```diff
@@ -0,0 +1,199 @@
+            single_file = True,
+            cfg = "target",
+        ),
+        "original_image": attr.string(
+            doc = "name of original image (for computing sha)",
+            mandatory = True,
+        ),
+        "image_name": attr.string(
+            doc = "name of image to run commands on",
+            mandatory = True,
+        ),
+        "command": attr.string_list(
+            doc = "command to run",
+            mandatory = True,
+            non_empty = True,
+        ),
+        "_run_tpl": attr.label(
+            default = Label("//util:commit.sh.tpl"),
+            allow_files = True,
+            single_file = True,
+        ),
+    },
+    executable = True,
+    implementation = _commit_impl,
+)
+
+_run_and_extract = rule(
+    attrs = {
+        "flags": attr.string_list(
+            doc = "list of flags to pass to run command",
+            default = [],
+        ),
+        "image_tar": attr.label(
+            executable = True,
+            allow_files = True,
+            mandatory = True,
+            single_file = True,
+            cfg = "target",
+        ),
+        "image_name": attr.string(
+            doc = "name of image to run commands on",
+            mandatory = True,
+        ),
+        "command": attr.string_list(
+            doc = "command to run",
+            mandatory = True,
+            non_empty = True,
+        ),
+        "extract_file": attr.string(
+            doc = "path to file to extract from container",
+            mandatory = True,
+        ),
+        "target": attr.string(
+            doc = "path to extract file to on host machine",
+            default = ".",
+        ),
+        "_extract_tpl": attr.label(
+            default = Label("//util:extract.sh.tpl"),
+            allow_files = True,
+            single_file = True,
+        ),
+    },
+    executable = True,
+    implementation = _extract_impl,
+)
+
+def container_run_and_commit(name, image, command, flags=None):
+    image_tar, intermediate_image = _rename_image(image, name)
+
+    _run_and_commit(
+        name = name,
+        original_image = image,
+        image_name = intermediate_image,
+        image_tar = image_tar + ".tar",
+        flags = flags,
+        command = command,
+    )
+
+def container_run_and_extract(name, image, command, extract_file, target, flags=None):
+    image_tar, intermediate_image = _rename_image(image, name)
+
+    _run_and_extract(
+        name = name,
+        image_name = intermediate_image,
+        image_tar = image_tar + ".tar",
+        flags = flags,
+        command = command,
+        extract_file = extract_file,
+        target = target,
+    )
+
+def _process_command(command_list):
+    # Use the $ to allow escape characters in string
+    return 'sh -c $\"{0}\"'.format(" && ".join(command_list))
+
+def _rename_image(image, name):
+    """A macro to predictably rename the image under test before threading
+    it to the container test rule."""
```

Comment:
Change the description to omit container test rule.

Also, we have been avoiding this by using the following logic to predict the image name.
I am not sure what issues you were facing..
https://github.com/GoogleCloudPlatform/base-images-docker/blob/master/package_managers/download_pkgs.bzl#L22.

Reply:
ah yeah, I copied this from the code in rules_docker. similar to what I mentioned above, it might be the case in this specific scenario but the image we're dealing with wasn't necessary always built with bazel, so inferring the name is sometimes impossible. I'll be moving this code to a common location in rules_docker so we can reuse it in a later PR.